### PR TITLE
[ci2] concurrency improvements

### PR DESCRIPTION
**Pull request by cseed**
- All subprocess calls are async.  The UI is much more responsive now.
 - Make refresh (rather than heal) non-reentrant.  There was a race condition where we could update the Github state of a PR we were actively deploying.  This seemed error prone.
 - We need to lock the repos until the build is fully deployed.  I now protect pr.heal.
 - Set `batch_changed = True` whenever heal needs to be rerun becomes some of its inputs (build_state, source or target sha, collection of PRs) changed.

Next up: deploy.

### CODE REVIEW

File: ci2/ci/github.py [hunk 1]
```diff
@@ -271,14 +274,42 @@ def __init__(self, branch):
...
     async def update(self, app):
+        # update everything
+        self.github_changed = True
+        self.batch_changed = True
+        await self._update(app)
+
+    async def _update(self, app):
+        if self.updating:
+            return
+        self.updating = True
```
Comment (danking) [inline ci2/ci/github.py]: shouldn't this be an atomically acquired lock (and if it's not available we just return)? Right now two concurrent threads of control could try cancelling the same batch job (which I guess is fine? but will trigger 400s in heal)

Comment (danking) [reply ci2/ci/github.py]: and maybe that supersedes the repos_lock because only one thread is updating at a time?

File: ci2/ci/github.py [hunk 2]
```diff
@@ -257,7 +259,8 @@ def batch_was_cancelled(batch):
             if len(batches) > 0:
                 self.batch = batches[0]
             elif run:
-                await self.start_build(batch)
+                async with repos_lock:
+                    await self.start_build(batch_client)
 
         if self.batch:
             seen_batch_ids.add(self.batch.id)
```
Comment (danking) [inline ci2/ci/github.py]: down here we change build_state but don't set batch_changed, that seems wrong

Comment (danking) [reply ci2/ci/github.py]: if a build is now success, we might want to merge it, right? batch_changed will trigger the watched branch heal loop that called this PR heal to try healing again and maybe merging.

File: ci2/ci/github.py (hunk 1 above)
Comment (cseed) [reply ci2/ci/github.py]: There should only be one thread, so this is atomic.

Comment (cseed) [reply ci2/ci/github.py]: Since there is no `await`.  Clarify of asyncio for the win.

**Comment by cseed** [comment]
> I don't understand why we need to use lock all start_builds.

start_build reads the repo (in particular, run_image command/script comes from the build.yaml), so another build can't start in the repo until the current one is done.

File: ci2/ci/github.py (hunk 2 above)
Comment (cseed) [reply ci2/ci/github.py]: Good catch, you're right. Fixed.

**Pull request merged**